### PR TITLE
eth: core: implement final block

### PR DESCRIPTION
**Pull request by MariusVanDerWijden**
This PR implements the notion of `FinalizedBlock` which is set post merge and all RPC calls with blocknumbers can now be called with `finalized` too

### CODE REVIEW

File: core/blockchain.go
Comment (holiman) [inline core/blockchain.go:508]: Maybe it doesn't matter (?)... 
But this is not concurrency-safe. If we get two `SetFinalized` calls, for `B1` and `B2` in quick succession, while the db is busy writing some large batch, then 
1. We set the atomic to B1, 
2. We get blocked on writing B1 to disk, 
3. We set the atomic to B2
4. We get blocked on writing B2 to disk
5. At some point later, B1 and B2 are written to disk, order not guaranteed. 

So there may be a mismatch between in-memory marker and disk-marker. But as I said, this may not be an issue at all (we seem to rely on the disk only for bootstrap, and use in-memory marker during normal operations)

**Review by holiman — APPROVED** [review]
One open question, but I suspect it's of no consequence, so LGTM

**Comment by MariusVanDerWijden** [comment]
Something doesn't work here, need to test it on kiln

File: core/blockchain.go
Comment (MariusVanDerWijden) [reply core/blockchain.go:508]: Hmm it means you could roll back finalization a couple of blocks for the user if a node restarts.
I don't think its a big issue, but would love to get @karalabe's opinion

**Comment by MariusVanDerWijden** [comment]
We've decided to add a lock to the FCU to make sure that all calls to it are correctly synchronized.
It doesn't have a big impact on performance but makes it way easier to reason about correctness of the method

File: core/blockchain.go
Comment (karalabe) [inline core/blockchain.go:488]: Wondering if it makes sense to add the TD, or rather to remove it from all other logs once PoS switchover hits.

Comment (MariusVanDerWijden) [reply core/blockchain.go:488]: Hmm maybe just remove all of them in a big swoop after the merge?

**Review by holiman — APPROVED** [review]
LGTM

**Comment by karalabe** [comment]
So, I guess one peculiarity about this is that if there's a reorg (e.g. crash + rewind), then we might end up with a non-existing (i.e. rewound) finalized block. That might get weird. Should we unset the finalized block to `nil` if there's a reorg/rewind?

**Comment by MariusVanDerWijden** [comment]
Hmmm, I don't know. 
Kinda depends on what the user wants to do with the finalized block.
If they only care about which blockheight was finalized, they would prefer us to not set it to nil.
If they want to query something based on it, or run some eth_call or something, than setting it to nil would be safer

File: core/rawdb/accessors_chain.go
Comment (karalabe) [inline core/rawdb/accessors_chain.go:231]: `fast` -> `finalized`

File: eth/catalyst/api.go
Comment (karalabe) [inline eth/catalyst/api.go:64]: No abbreviations, please.

File: eth/catalyst/api.go
Comment (karalabe) [inline eth/catalyst/api.go:65]: Also, if it's only for the fork choice, please name to `forkChoiceLock`

File: internal/jsre/deps/web3.js
Comment (karalabe) [inline internal/jsre/deps/web3.js:3699]: You need to regenerate the bundled file, otherwise this change is moot.

File: eth/catalyst/api.go
Comment (karalabe) [inline eth/catalyst/api.go:65]: Ah wait, this should not be a pointer.

File: eth/catalyst/api.go
Comment (karalabe) [inline eth/catalyst/api.go:78]: You don't need this, just don't make it a pointer.

**Review by karalabe — APPROVED** [review]
SGTM

File: eth/api_backend.go
Comment (karalabe) [inline eth/api_backend.go:111]: No need for the `else`, keeps the code cleaner.

File: eth/api_backend.go
Comment (karalabe) [inline eth/api_backend.go:75]: No need for the `else`, keeps the code cleaner.

**Review by karalabe — APPROVED** [review]
SGTM

**Comment by 0903482095** [comment]
when I call method eth_getBlockByNumber with the tag "finalized", **this block will not be chain-reorg, right?**
(because I'm using the tag "latest" and need the check chain-reorg)

**Pull request merged**
